lib/maths/CDataFrameUtils.cc

```diff
@@ -57,6 +61,162 @@ bool doReduce(std::pair<std::vector<READER>, bool> readResults,
+    double numberTrainingRows{allTrainingRowsMask.manhattan()};
+    std::size_t desiredCount{
+        (static_cast<std::size_t>(numberTrainingRows) + numberFolds / 2) / numberFolds};
+    CSampling::weightedSample(desiredCount, frequencies, categoryCounts);
+    LOG_TRACE(<< "desired category counts per test fold = "
+              << core::CContainerPrinter::print(categoryCounts));
+
+    auto sampler = std::make_unique<CStratifiedSampler>(categoryCounts.size());
+    for (std::size_t i = 0; i < categoryCounts.size(); ++i) {
+        sampler->addSampler(categoryCounts[i], rng);
+    }
+    sampler->samplerSelector([targetColumn](const TRowRef& row) mutable {
+        return static_cast<std::size_t>(row[targetColumn]);
+    });
+
+    return {std::move(sampler), std::move(frequencies)};
+}
+
+//! Get a regression stratified row sampler for cross fold validation.
+std::unique_ptr<CStratifiedSampler>
+regressionStratifiedCrossValiationRowSampler(std::size_t numberThreads,
+                                             const core::CDataFrame& frame,
+                                             std::size_t targetColumn,
+                                             CPRNG::CXorOShiro128Plus rng,
+                                             std::size_t numberFolds,
+                                             const core::CPackedBitVector& allTrainingRowsMask) {
+
+    CDataFrameUtils::TQuantileSketchVec quantiles;
+    CDataFrameUtils::columnQuantiles(
+        numberThreads, frame, allTrainingRowsMask, {targetColumn},
+        CQuantileSketch{CQuantileSketch::E_Linear, 50}, quantiles);
+
+    TDoubleVec buckets;
+    for (double percentile = 10.0; percentile < 100.0; percentile += 10.0) {
+        double xDecile;
+        quantiles[0].quantile(percentile, xDecile);
+        buckets.push_back(xDecile);
+    }
+    buckets.erase(std::unique(buckets.begin(), buckets.end()), buckets.end());
+    buckets.push_back(std::numeric_limits<double>::max());
+    LOG_TRACE(<< "buckets = " << core::CContainerPrinter::print(buckets));
+
+    auto bucketSelector = [buckets, targetColumn](const TRowRef& row) mutable {
+        return static_cast<std::size_t>(
+            std::upper_bound(buckets.begin(), buckets.end(), row[targetColumn]) -
+            buckets.begin());
+    };
+
+    auto countBucketRows = core::bindRetrievableState(
+        [&](TDoubleVec& bucketCounts, TRowItr beginRows, TRowItr endRows) {
+            for (auto row = beginRows; row != endRows; ++row) {
+                bucketCounts[bucketSelector(*row)] += 1.0;
+            }
+        },
+        TDoubleVec(buckets.size(), 0.0));
+    auto copyBucketRowCounts = [](TDoubleVec counts_, TDoubleVec& counts) {
+        counts = std::move(counts_);
+    };
+    auto reduceBucketRowCounts = [](TDoubleVec counts_, TDoubleVec& counts) {
+        for (std::size_t i = 0; i < counts.size(); ++i) {
+            counts[i] += counts_[i];
+        }
+    };
+
+    TDoubleVec weights;
```

Comment:
The variable name `weights` is confusing here. Can you maybe rename it to something more self-explanatory, i.e. `decentileCounts`?

Reply:
Ok good suggestion. I renamed in [this](https://github.com/elastic/ml-cpp/pull/784/commits/7c23e983bf26fe663ffc982ae2adf4c18a7949da) commit. (Note I don't use decile because I passed in the number of buckets to optionally disable.)